### PR TITLE
Limit kubelets from updating their own labels when NodeRestriction is enabled

### PR DESCRIPTION
**Pull request by liggitt**
Implements phase 1 of https://github.com/kubernetes/community/blob/master/keps/sig-auth/0000-20170814-bounding-self-labeling-kubelets.md#implementation-timeline

Docs PR in https://github.com/kubernetes/website/pull/10944

This PR:
* Prevents kubelets from adding/modifying labels prefixed with `node-restriction.kubernetes.io` on their Node objects
* Restricts which `kubernetes.io` labels kubelets can set when *updating* their Node objects to an allowed set of labels and label prefixes
* Warns on kubelet startup if `kubernetes.io` labels outside that set are passed via `--node-labels` (will escalate to an error in v1.15)
* Warns in admission if `kubernetes.io` labels outside that set are passed on Node creation (will escalate to a forbidden error in v1.17)

/sig auth
/sig node
/sig storage

/cc @mikedanese @verult @saad-ali @vishh 

```release-note
kube-apiserver: the `NodeRestriction` admission plugin now prevents kubelets from modifying `Node` labels prefixed with `node-restriction.kubernetes.io/`. The `node-restriction.kubernetes.io/` label prefix is reserved for cluster administrators to use for labeling `Node` objects to target workloads to nodes in a way that kubelets cannot modify or spoof.

kubelet: it is now deprecated to use the `--node-labels` flag to set `kubernetes.io/` and `k8s.io/`-prefixed labels other than the following labels:
* `kubernetes.io/hostname`
* `kubernetes.io/instance-type`
* `kubernetes.io/os`
* `kubernetes.io/arch`
* `beta.kubernetes.io/instance-type`
* `beta.kubernetes.io/os`
* `beta.kubernetes.io/arch`
* `failure-domain.kubernetes.io/zone`
* `failure-domain.kubernetes.io/region`
* `failure-domain.beta.kubernetes.io/zone`
* `failure-domain.beta.kubernetes.io/region`
* `[*.]kubelet.kubernetes.io/*`
* `[*.]node.kubernetes.io/*`
Setting other `kubernetes.io/`- and `k8s.io/`-prefixed labels using the `--node-labels` flag will produce a warning in v1.13, and be disallowed in v1.15. Setting labels that are not prefixed with `kubernetes.io/` or `k8s.io/` is still permitted.
```

### CODE REVIEW

**Comment by liggitt** [comment]
cc @vikaschoudhary16 @derekwaynecarr since the [plugin watcher](https://github.com/kubernetes/kubernetes/pull/67684/commits/b0c3be7ef25197486d89cc7ec72ab6e3f712d98f#r215129853) looks to be the most immediate source of expanding kubelet self-labeling. is that attempting to go to beta in 1.12?

**Comment by vishh** [comment]
@liggitt 

> Use cases like device plugins and CSI drivers intend to use self-added labels to express topology, which is fine, but those use cases need to be coordinated with controls the cluster-admin can use to ensure labels remain useful for partitioning a cluster along security-related boundaries

Can you explain this more? I want to make sure usability isn't compromised a lot. Poking a hole in RBAC isn't safe today for node level daemons.

**Comment by liggitt** [comment]
> Can you explain this more? I want to make sure usability isn't compromised a lot. Poking a hole in RBAC isn't safe today for node level daemons.

For things the kubelet is expected to self-report topology for (device drivers and CSI come to mind), one possibility would be to have a registration object for the device class or CSI driver that includes the list of label keys that will be used for topology. For example, CSI has the `CSIDriver` object which would be a natural place to put that info (and has benefits beyond authorization... today the CSI provisioner has to pick a node at random to determine what the topology keys are).

That registration object would not be kubelet-writeable. The node admission plugin could then allow kubelets to self-set labels that were designated as topology labels by one of those registration objects.

**Comment by vishh** [comment]
On Fri, Oct 5, 2018 at 8:14 AM Jordan Liggitt <notifications@github.com>
wrote:

> Can you explain this more? I want to make sure usability isn't compromised
> a lot. Poking a hole in RBAC isn't safe today for node level daemons.
>
> For things the kubelet is expected to self-report topology for (device
> drivers and CSI come to mind), one possibility would be to have a
> registration object for the device class or CSI driver that includes the
> list of label keys that will be used for topology. For example, CSI has the
> CSIDriver object which would be a natural place to put that info (and has
> benefits beyond authorization... today the CSI provisioner has to pick a
> node at random to determine what the topology keys are).
>
> That registration object would not be kubelet-writeable. The node
> admission plugin could then allow kubelets to self-set labels that were
> designated as topology labels by one of those registration objects.
>
How is auth managed for the plugins that write this secondary object which
will then get copied over to the Node Object?

**Comment by liggitt** [comment]
> How is auth managed for the plugins that write this secondary object which will then get copied over to the Node Object?

Either the cluster admin or something at the control plane level would be expected to set those objects up (it could be part of the manifest or add-on that set up the device driver or CSI plugin)

**Comment by liggitt** [comment]
cc @kubernetes/sig-auth-pr-reviews @kubernetes/sig-node-pr-reviews @kubernetes/sig-storage-pr-reviews

File: plugin/pkg/admission/noderestriction/admission_test.go [hunk 1]
```diff
@@ -740,6 +744,12 @@ func Test_nodePlugin_Admit(t *testing.T) {
 			attributes: admission.NewAttributesRecord(mynodeObjTaintA, nil, nodeKind, mynodeObj.Namespace, "", nodeResource, "", admission.Create, false, mynode),
 			err:        "",
 		},
+		{
+			name:       "allow create of my node with labels",
+			podsGetter: noExistingPods,
+			attributes: admission.NewAttributesRecord(mynodeObjLabelA, nil, nodeKind, mynodeObj.Namespace, "", nodeResource, "", admission.Create, false, mynode),
```
Comment (mikedanese) [inline plugin/pkg/admission/noderestriction/admission_test.go]: s/mynodeObj.Namespace/mynodeObjLabelA.Namespace/ here and below? Or just omit it. We are already omitting the name and nodes aren't namespaced. It looks like a number of other test cases are sloppy as well.

**Review by mikedanese — COMMENTED** [review]
Looks reasonable. Who needs to sign off on this?

File: plugin/pkg/admission/noderestriction/admission_test.go (hunk 1 above)
Comment (liggitt) [reply plugin/pkg/admission/noderestriction/admission_test.go]: it doesn't actually bother me to pass the empty object namespace in, and the name of all the `mynode*`test fixtures is identical, the point of the different ones is to change other fields in the object. will address separately if desired, trying to keep the test diff readable.

File: plugin/pkg/admission/noderestriction/admission.go [hunk 2]
```diff
@@ -37,24 +38,38 @@ import (
...
+		kubeletapis.LabelOS,
+		kubeletapis.LabelArch,
+	)
+}
+
 // Register registers a plugin
 func Register(plugins *admission.Plugins) {
 	plugins.Register(PluginName, func(config io.Reader) (admission.Interface, error) {
-		return NewPlugin(nodeidentifier.NewDefaultNodeIdentifier()), nil
+		// TODO: read allowed labels from config
```
[the remark this answers is not in the document]
Comment (liggitt) [reply plugin/pkg/admission/noderestriction/admission.go]: updated to implement https://github.com/kubernetes/community/pull/911, pinning to specific prefixes

**Comment by liggitt** [comment]
/hold
for https://github.com/kubernetes/community/pull/911

File: pkg/kubelet/apis/well_known_labels.go [hunk 3]
```diff
@@ -26,8 +31,65 @@ const (
...
+}
+
+var allowedLabelNamespaces = map[string]bool{
+	LabelNamespaceSuffixKubelet:  true,
+	LabelNamespaceSuffixTopology: true,
+}
+
+// AllowedLabels returns the list of label keys kubelets are allowed to set on their own Node objects
+func AllowedLabels() []string {
+	retval := make([]string, 0, len(allowedLabels))
```
Comment (enj) [inline pkg/kubelet/apis/well_known_labels.go]: `return StringKeySet(allowedLabels).List()`?

File: pkg/kubelet/apis/well_known_labels.go [hunk 4]
```diff
@@ -26,8 +31,65 @@ const (
...
+	for s := range allowedLabels {
+		retval = append(retval, s)
+	}
+	sort.Strings(retval)
+	return retval
+}
+
+// AllowedLabelNamespaces returns the list of label key namespaces kubelets are allowed to set on their own Node objects
+func AllowedLabelNamespaces() []string {
+	retval := make([]string, 0, len(allowedLabelNamespaces))
```
Comment (enj) [inline pkg/kubelet/apis/well_known_labels.go]: `return StringKeySet(allowedLabelNamespaces).List()`?

**Review by enj — COMMENTED** [review]
Minor comments, LGTM

**Comment by liggitt** [comment]
@mikedanese ready for review

**Review by mikedanese — COMMENTED** [review]
still lgtm, had a few minor comments.

File: pkg/kubelet/apis/well_known_labels.go [hunk 5]
```diff
@@ -26,8 +32,85 @@ const (
...
+	return kubeletLabels.List()
+}
+
+// KubeletLabelNamespaces returns the list of label key namespaces kubelets are allowed to set on their own Node objects
+func KubeletLabelNamespaces() []string {
+	return kubeletLabelNamespaces.List()
+}
+
+// IsKubernetesLabel returns true if the given label key is under the kubernetes.io or k8s.io namespace
+func IsKubernetesLabel(key string) bool {
```
[the remark this answers is not in the document]
Comment (liggitt) [reply pkg/kubelet/apis/well_known_labels.go]: eh, I guess not. moved to private function where used

File: plugin/pkg/admission/noderestriction/admission_test.go [hunk 6]
```diff
@@ -817,6 +924,24 @@ func Test_nodePlugin_Admit(t *testing.T) {
 			attributes: admission.NewAttributesRecord(mynodeObjTaintA, mynodeObjTaintA, nodeKind, mynodeObj.Namespace, mynodeObj.Name, nodeResource, "", admission.Update, false, mynode),
 			err:        "",
 		},
+		{
+			name:       "allow update of my node: add allowed labels",
```
[the remark this answers is not in the document]
Comment (liggitt) [reply plugin/pkg/admission/noderestriction/admission_test.go]: added these test cases:
* allow update of my node: add allowed labels while forbidden labels exist unmodified
* allow update of my node: remove allowed labels while forbidden labels exist unmodified

File: plugin/pkg/admission/noderestriction/admission_test.go [hunk 7]
```diff
@@ -817,6 +924,24 @@ func Test_nodePlugin_Admit(t *testing.T) {
...
+			podsGetter: existingPods,
+			attributes: admission.NewAttributesRecord(setAllowedUpdateLabels(mynodeObj, ""), mynodeObj, nodeKind, mynodeObj.Namespace, mynodeObj.Name, nodeResource, "", admission.Update, false, mynode),
+			err:        "",
+		},
+		{
+			name:       "allow update of my node: remove allowed labels",
```
[the remark this answers is not in the document]
Comment (liggitt) [reply plugin/pkg/admission/noderestriction/admission_test.go]: added test case:
* allow update of my node: modify allowed labels

**Comment by liggitt** [comment]
comments addressed

**Comment by derekwaynecarr** [comment]
kubelet changes look good to me.

/lgtm
/approve

**Comment by kfox1111** [comment]
Saw this in the release notes. This is very good to see.

One thing I don't understand though. To really solve the issue, all usage of labels that could ever have security ramifications now would be required to be prefixed with kubernetes.io/ or k8s.io/ 
But most of the time its user workload. So we are encouraging users to come up with their own labels under the kubernetes.io/ or k8s.io/  spaces?

Thanks,
Kevin

**Comment by liggitt** [comment]
> all usage of labels that could ever have security ramifications now would be required to be prefixed with kubernetes.io/ or k8s.io/
> But most of the time its user workload. So we are encouraging users to come up with their own labels under the kubernetes.io/ or k8s.io/ spaces?

See https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#noderestriction for a label prefix specifically reserved for this use:

> Prevents kubelets from adding/removing/updating labels with a `node-restriction.kubernetes.io/` prefix. This label prefix is reserved for administrators to label their Node objects for workload isolation purposes, and kubelets will not be allowed to modify labels with that prefix.

**Comment by kfox1111** [comment]
Ah. ok. So, the recommendation is in general, "always prefix your node labels with node-restriction.kubernetes.io/"?

**Comment by kfox1111** [comment]
Ok. Yeah, was just looking for general advice to give to new users to avoid issues. Since its always safe, and doesn't have an extra cost, its "always prefix your node labels with node-restriction.kubernetes.io/" (unless you know what you are doing). Thanks. :)

**Comment by ravilr** [comment]
@liggitt  'node-role.kubernetes.io/master' and 'node-role.kubernetes.io/node' seems to be used across  projects. Can you please give some direction on what do you think about whitelisting this as allowed set of labels ? 
https://github.com/kubernetes/kubernetes/blob/954996e231074dc7429f7be1256a579bedd8344c/pkg/controller/service/service_controller.go#L66
Also across kubeadm, kubespray

**Comment by pires** [comment]
Conformance tests will also be broken if this annotation is not used for control-plane nodes and they're unschedulable, eg with taints.

**Comment by liggitt** [comment]
that label is not appropriate for use in a conformance test, as it is not official API and is not maintained by kubernetes components

**Comment by liggitt** [comment]
opened https://github.com/kubernetes/kubernetes/pull/76654 to let the e2e/conformance invoker select which nodes they think should be scheduleable, rather than hard-code handling of a label

**Comment by liggitt** [comment]
> Can you please give some direction on what do you think about whitelisting this as allowed set of labels ?

Letting aribtrary nodes self-label as a master node is not reasonable. That label is likely to be used in daemonset nodeSelectors running highly-privileged components.

**Comment by liggitt** [comment]
kubeadm does not set that label using self-labeling, but applies it using the superuser client at cluster startup:

https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/phases/markcontrolplane/markcontrolplane.go#L28-L44

**Pull request merged**